### ISSUE TITLE
voila / ipyvuetify /ipydatagrid issues

### ISSUE THREAD

**Issue by joseberlines** [bloomberg/ipydatagrid#231]
I am running a notebook (Jupyter lab) in a Jupyter hub and I got the following situation:

In this case I imported ipydatagrid as such import ipydatagrid.
Ipydatagrid ended not being used in the whole notebook.
![image](https://user-images.githubusercontent.com/33203439/116075997-f4a83600-a693-11eb-82cb-d1240162b2d6.png)
Now as it is seen on the left when running the notebook the GUI (ipyvuetify) is rendered.
But on the right as can be seen voila DOES NOT RENDER THE IPYVUETIFY COMPONENTS.

Now if I comment out the import ipyvuetify line at the beginning then the result is that voila renders:
![image](https://user-images.githubusercontent.com/33203439/116076324-608a9e80-a694-11eb-99a1-2f4a3affca36.png)

This was after i did this:
![image](https://user-images.githubusercontent.com/33203439/116076411-7ac47c80-a694-11eb-9fe7-d1bc0b57867d.png)

**Comment by joseberlines** [comment]
I reproduced this again in the following way.
I downloaded the ipyvuetify repo into a folder (no particular venv)
run the Datagrid notebook example. All fine, the tables were shown.
I imported ipyvuetify and when doind so the button was not shown

![image](https://user-images.githubusercontent.com/33203439/116095451-b6694180-a6a8-11eb-9352-401b814c5025.png)

Actually I realise that this might be an issue for ipyvuetify....

**Comment by ibdafna** [comment]
@joseberlines I just tried running both in a new environment and everything seem to work fine:

![image](https://user-images.githubusercontent.com/24281433/116128695-10a0dd00-a67e-11eb-9dee-27c0cb22af1c.png)

This is the command I used to set up my environment:
```bash
conda create -n t9 -c conda-forge python=3.8 ipydatagrid ipyvuetify jupyterlab
```

**Comment by joseberlines** [comment]
Ok. And if you now try to render the notebook with voila?

**Comment by ibdafna** [comment]
Just tested - the issue with voila. It tries loading the front-end script separately from a CDN

**Comment by ibdafna** [comment]
As we JLab3-only compatible widgets such as `ipydatagrid` do not publish a separate package for the view (everything is bundled via the Python package manager), any attempt as loading non-existent `npm` packages will fail.

**Comment by joseberlines** [comment]
This goes beyond my code knowledge (I only know python) but perhaps @jtpio could have a look. In my examples voila is installed and just by clicking the button you can see the rendering in the right side.

**Comment by ibdafna** [comment]
@jtpio will definitely know more!

**Comment by joseberlines** [comment]
And in your example can you confirm that if you ONLY have one of the both (ipyvuetify OR ipydatagrid) voila would render both?

**Comment by joseberlines** [comment]
This is also interesting for @mariobuikhuizen perhaps

**Comment by ibdafna** [comment]
`ipyvuetify` on its own works because there is a separate `npm` package published for it, which `voila` can load from the CDN. `ipydatagrid` fails because such package does not exist. I would like to avoid publishing a separate package for the view unless it's absolutely necessary, so let's see how it goes with the `voila` folks supporting the new JLab3-compatible workflow.

**Comment by joseberlines** [comment]
@ibdafna thanks for trying it out

**Comment by SylvainCorlay** [comment]
@ibdafna I think that the JupyterLab plugin for ipydatagrid should be published on npm.

That will ensure that it works with nbconvert and voilà.

**Comment by marcmenem** [comment]
FWIW, `ipydatagrid` works correctly for me with `voila` in Firefox, but it is failing in Safari. I have so many `voila` related issues that I lost count...

**Comment by SylvainCorlay** [comment]
@marcmenem you probably are using the local nbextensions to get it to work on voilà since ipydatagrid is not published on npm.

**Comment by joseberlines** [comment]
Since I am not proficient in JS I can never relate the issues. Voila or the widget library.
As a use case that might be worth to mention. When working in a jupyter hub the idea is that many user are going to use applications via a url without even knowing what is a jupyter notebook. This is important to mention since often times there are suggestions as to trying something in the terminal. If it necessary to hack in the terminal things get restricted to a few user who even know what a terminal is. I am talking about several thousand of coworkers who only know pythons from zoos. ;-) 😉

**Comment by marcmenem** [comment]
> @marcmenem you probably are using the local nbextensions to get it to work on voilà since ipydatagrid is not published on npm.

Yes, it is indeed the case, but it is not clear to me why it does not work the same on all browsers. I think the issue has more to do with `voila` then `ipydatagrid`.

**Comment by joseberlines** [comment]
Wait a minute, 
for some reason I thought that the problem might be an interference between ipydatagrid and ipyvuetify, but actually ipydatagrids are not shown in jup Lab 3 (jupyterhub  1.2.2 jupyterlab  3.0.10) when clicking the button for rendering.
Since my JS knowledge is ver limited I dont know if this is a voila issue or a ipydatagrid one.
![image](https://user-images.githubusercontent.com/33203439/116368066-42de4600-a808-11eb-9a8f-c9886f36e17b.png)

**Comment by joseberlines** [comment]
Hi all again. Just to make clear here the implications. Working in a jupyter hub implies that notebooks can be shared with a link including    www.WHATEVERHUB.COM/voila/render/mypathtothenotebook.ipynb
Even in the event that the rendering might be possible from the terminal line with particular commands, this situation makes impossible to chare notebooks including ipydatagrids across an organization. We used notebooks wiht ipyvuetify to distribute across the organizatino to people who dont have any clue about coding/terminal/python...
I think that if ipydatagrids can be rendered in voila the power of this possibiity is inmense. Is like having excel, customized, powered with python

**Comment by ibdafna** [comment]
@joseberlines you should be able to use `ipydatagrid` in `voila` via the installed nbextension, should work with the following command: `voila --enable_nbextensions=True`. We will eventually publish an NPM package, too, but for now hopefully this will work.

**Comment by joseberlines** [comment]
Thanks @ibdafna . I will have a look to your suggestion and report back. That being said I think being able to render ipydatagrid via voila without touching the terminal hugely increments the potentiality of ipDatagrid since anyone Can serve solutions via a jupyter hub solution without the users having to touch the terminal, only sending a url. (Oftentimes users don't even know what a notebook is, those are ou of the equation when it comes to rendering via terminal commands.). Thanks a lot for all the work.

**Comment by ibdafna** [comment]
`ipydatagrid` 1.0.5 has been released. This version also has CDN support so it should render in `voila` as you'd expect. I just tested it locally and it worked. Are you able to test and let me know if it works on your end, please?

**Comment by joseberlines** [comment]
checked:
So cool. Ipydatagrid can be now part of applications distributed rendered with voila via a jupyter hub!!!!! 
Congratulations
![image](https://user-images.githubusercontent.com/33203439/117671825-81c9af80-b1a9-11eb-9cf5-3910c71ac77f.png)

**Comment by joseberlines** [comment]
should I close this issue?

**Comment by ibdafna** [comment]
Glad it's all working now 😸 Let us know if anything else come up. Will close this in the meantime!
